task-queue-multi-consumers/src/main/java/MultiConsumersTaskQueue.java

```diff
@@ -0,0 +1,135 @@
+        }
+        return consumerThreads;
+    }
+
+    @Override
+    public void runTask(Runnable r) {
+        if (closed)
+            throw new IllegalStateException("Can't add tasks to a closed queue");
+
+        synchronized (tasks) {
+            tasks.offer(r);
+        }
+
+        runQueueTasks();
+    }
+
+    @Override
+    public void close() {
+        closed = true;
+        boolean allocatedConsumer;
+
+        do {
+            allocatedConsumer = runQueueTasks();
+        } while (!allocatedConsumer);
+    }
+
+    private boolean runQueueTasks() {
+        boolean allocatedConsumer = false;
+        ConsumerThread consumerThread = null;
+
+        synchronized (availableConsumers) {
+            if (!availableConsumers.isEmpty()) {
+                consumerThread = availableConsumers.iterator().next();
+                availableConsumers.remove(consumerThread);
+                allocatedConsumer = true;
+            }
+        }
+
+        if (consumerThread != null) {
+            synchronized (busyConsumers) {
+                busyConsumers.add(consumerThread);
+            }
+            consumerThread.consumeTasks();
+        }
+        return allocatedConsumer;
+    }
+
+    private void notifyDone(ConsumerThread consumerThread) {
+        synchronized (busyConsumers) {
+            busyConsumers.remove(consumerThread);
+        }
+
+        synchronized (availableConsumers) {
+            availableConsumers.add(consumerThread);
+        }
+    }
+
+    //for testing
+    Queue<Runnable> getQueue() {
+        return tasks;
+    }
+
+    private class ConsumerThread implements Runnable {
+        private final String id;
+        private Runnable task;
+        private Thread consumerThread;
+
+        ConsumerThread(String id) {
+            this.id = id;
+            task = null;
+            consumerThread = null;
+        }
+
+        private void consumeTasks() {
+            if (consumerThread == null) {
+                consumerThread = new Thread(this);
+                consumerThread.start();
+            }
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                synchronized (tasks) {
+                    if (!tasks.isEmpty()) {
+                        task = tasks.poll();
+                    } else break;
+                }
+                System.out.println("Running with thread id = " + id);
+                task.run();
+            }
+            consumerThread = null;
+            notifyDone(this);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ConsumerThread that = (ConsumerThread) o;
+            return Objects.equals(id, that.id);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(id);
+        }
+    }
+}
```

Comment:
I finished reviewing this class, and here are my comments:
1. nice work
2. as I wrote earlier, I think it's better not to rely on such specific features of Java. it's technically valid, and if we would be writing the JDK itself then perhaps we'd use these features, but it's not the case.
3. you rely on the state of both "busy" and "available" data structures, but when you read/modify them, you only lock one of them at a time. this would eventually lead to a race condition. instead, use a single lock object (`private Object myLock = new Object();`).